### ISSUE TITLE
HEAD requests returning non-empty entity body

### ISSUE THREAD

**Issue by bnordbo** [yesodweb/wai#369]
I suspect there could be an issue in `warp 3.0.10.1` and onwards with regard to how responses to `HEAD` requests are handled. For instance, consider this server:

```haskell
main :: IO ()
main = run 1234 $ \_ rsp -> rsp $ responseLBS status200 [] empty
```

When issuing the following HTTP request:

```
echo -ne "HEAD / HTTP/1.1\r\nHost: localhost\r\n\r\n" | nc localhost 1234
```

The server will respond as follows:

```
HTTP/1.1 200 OK
Transfer-Encoding: chunked
Date: Wed, 22 Apr 2015 10:09:03 GMT
Server: Warp/3.0.10.1

0

```

As far as I can see (RFC 2616 [4.3](http://tools.ietf.org/html/rfc2616#section-4.3) and [9.4](http://tools.ietf.org/html/rfc2616#section-9.4)), the last-chunk marker should not be present in this case. If used from an application that re-uses connections, the superfluous `0` will typically be kept in the receive buffer, and trip up any subsequent request with an `InvalidStatusLine "0"` error.

It seems this issue was introduced by 699b9f823b7e8ad39c353292969eddeb1d705b87 in relation to ensuring that the response headers for a `HEAD` request matches the ones for a `GET`.

**Comment by snoyberg** [comment]
Pinging @kazu-yamamoto, can you have a look at this?

**Comment by kazu-yamamoto** [comment]
The patch above solves this problem.

@snoyberg Would you review this patch?

**Comment by snoyberg** [comment]
Looks good to me. @bnordbo can you confirm?

**Comment by bnordbo** [comment]
@snoyberg, @kazu-yamamoto, yes, it looks good. Thanks for the quick response!

**Comment by bnordbo** [comment]
I also applied the patch, and my test case passes now. I notice that it also causes the `Transfer-Encoding` header to be removed from the `HEAD` response. This is permitted according to RFC 2616, section 9.4, which states that:

> The metainformation contained in the HTTP headers in response to a HEAD request SHOULD be identical to the information sent in response to a GET request.

**Comment by snoyberg** [comment]
Awesome, thanks for the report and testing, and thanks for fixing @kazu-yamamoto. I've released to Hackage as 3.0.12.1.
